rules/terraformrules/terraform_unused_declaration.go

```diff
@@ -0,0 +1,169 @@
+package terraformrules
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/hashicorp/terraform/configs"
+	"github.com/terraform-linters/tflint/tflint"
+)
+
+// TerraformUnusedDeclarationsRule checks whether variables, data sources, or locals are declared but unused
+type TerraformUnusedDeclarationsRule struct{}
+
+type declarations struct {
+	Variables     map[string]*configs.Variable
+	DataResources map[string]*configs.Resource
+	Locals        map[string]*configs.Local
+}
+
+// NewTerraformUnusedDeclarationsRule returns a new rule
+func NewTerraformUnusedDeclarationsRule() *TerraformUnusedDeclarationsRule {
+	return &TerraformUnusedDeclarationsRule{}
+}
+
+// Name returns the rule name
+func (r *TerraformUnusedDeclarationsRule) Name() string {
+	return "terraform_unused_declarations"
+}
+
+// Enabled returns whether the rule is enabled by default
+func (r *TerraformUnusedDeclarationsRule) Enabled() bool {
+	return false
+}
+
+// Severity returns the rule severity
+func (r *TerraformUnusedDeclarationsRule) Severity() string {
+	return tflint.WARNING
+}
+
+// Link returns the rule reference link
+func (r *TerraformUnusedDeclarationsRule) Link() string {
+	return tflint.ReferenceLink(r.Name())
+}
+
+// Check emits issues for any variables, locals, and data sources that are declared but not used
+func (r *TerraformUnusedDeclarationsRule) Check(runner *tflint.Runner) error {
+	log.Printf("[TRACE] Check `%s` rule for `%s` runner", r.Name(), runner.TFConfigPath())
+
+	decl := r.declarations(runner.TFConfig.Module)
+	for _, resource := range runner.TFConfig.Module.ManagedResources {
+		r.checkForVariablesInBody(runner, resource.Config, decl)
+	}
+	for _, data := range runner.TFConfig.Module.DataResources {
+		r.checkForVariablesInBody(runner, data.Config, decl)
+	}
+	for _, provider := range runner.TFConfig.Module.ProviderConfigs {
+		r.checkForVariablesInBody(runner, provider.Config, decl)
+	}
+	for _, module := range runner.TFConfig.Module.ModuleCalls {
+		r.checkForVariablesInBody(runner, module.Config, decl)
+	}
+	for _, output := range runner.TFConfig.Module.Outputs {
+		r.checkForVariablesInExpr(runner, output.Expr, decl)
+	}
+	for _, local := range runner.TFConfig.Module.Locals {
+		r.checkForVariablesInExpr(runner, local.Expr, decl)
+	}
```

Comment:
Implementing functions like `WalkExpressions` in the `Runner` can simplify the implementation of rules that visit all expressions. Imagine something like `WalkResourceAttributes`.

https://github.com/terraform-linters/tflint/blob/689426718cd1d2c499f8de996fbcf06e1533e387/tflint/runner.go#L505-L521

It's a refactoring idea, so you can add all the rules and then work on another pull request 👍